### PR TITLE
Create Listing - Support 721

### PR DESCRIPTION
**Pull request by LukasJenicek**
- Use viem library to support typesafety while calling ERC contracts and marketplace Contract
- Support ERC1155 and ERC721
- Use marketplace_v2 as default marketplace ( could be overwritten using `--marketplace-version "v1"`

Marketplace Example V2: 
```
./dist/bin.js marketplace create-listings  \ 
   --key "0xxxxxxxxxxxxxxxx" \ 
   --address "0x602d5dc17490794267c7fa5f58a453eb9159a86d" \ 
   --token 8 \ 
   --network "polygon" \ 
   --currency "0x3c499c542cEF5E3811e1192ce70d8cC03d5c3359" \
   --price "1" \
   --type "ERC721"
```

Marketplace Example V1: 
```
./dist/bin.js marketplace create-listings  \ 
   --key "0xxxxxxxxxxxxxxxx" \ 
   --address "0x602d5dc17490794267c7fa5f58a453eb9159a86d" \ 
   --token 8 \ 
   --network "polygon" \ 
   --currency "0x3c499c542cEF5E3811e1192ce70d8cC03d5c3359" \
   --price "1" \
   --type "ERC721" \ 
   --marketplace-version "v1"
```

image

### CODE REVIEW

File: .prettierrc [hunk 1]
```diff
@@ -1,3 +1,8 @@
 {
-  "tabWidth": 4
-}
+  "arrowParens": "avoid",
+  "semi": false,
```
Comment (AlexanderKolberg) [inline .prettierrc]: We do this in a few projects, but I really hate it.. This is abusing js error correction, and the edge cases where prettier does it wrong is a time sink

File: src/marketplace/create_listings.ts [hunk 2]
```diff
@@ -1,196 +1,247 @@
...
+import { ethers, Numeric } from 'ethers'
+import { findSupportedNetwork } from '@0xsequence/network'
 import { isValidPrivateKey } from '../utils/'
-import { ERC1155_ABI } from "../abi/ERC_1155";
-import { Orderbook_ABI } from "../abi/Orderbook";
-
-
-const SEQUENCE_MARKETPLACE_V1_ADDRESS = "0xB537a160472183f2150d42EB1c3DD6684A55f74c";
+import { ERC1155_ABI } from '../abi/ERC_1155'
+import { Orderbook_ABI } from '../abi/Orderbook'
```
Comment (AlexanderKolberg) [inline src/marketplace/create_listings.ts]: Lets rename this to SequenceMarktetplace_V1_ABI

File: src/marketplace/create_listings.ts [hunk 3]
```diff
@@ -1,196 +1,247 @@
...
+    console.log(`Using node URL: ${nodeUrl}`)
+  }
+
+  const provider = new ethers.JsonRpcProvider(nodeUrl)
+  const wallet = new ethers.Wallet(privateKey, provider)
+
+  // Set up the wallet client
+  const walletClient = createWalletClient({
+    account: privateKeyToAccount(privateKey),
+    chain: polygon,
```
Comment (AlexanderKolberg) [inline src/marketplace/create_listings.ts]: This is polygon only... Lets use the chain config above

Very untested but I think you can get away with something like this:

```suggestion
    chain:  {
			...chainConfig,
			rpcUrls: {
				default: {
					http: [nodeUrl],
				}
			}
		}
```

File: src/marketplace/create_listings.ts [hunk 4]
```diff
@@ -1,196 +1,247 @@
...
+    transport: http(nodeUrl), // Replace with your Polygon node URL
```
Comment (AlexanderKolberg) [inline src/marketplace/create_listings.ts]: Remove comment

File: src/marketplace/create_listings.ts [hunk 5]
```diff
@@ -1,196 +1,247 @@
...
-  
-    const sequenceMarketplaceInterface = new ethers.Interface(Orderbook_ABI);
-  
-    const dataCreateRequest = sequenceMarketplaceInterface.encodeFunctionData(
-      "createRequest",
+  }
+
+  const marketplaceContract = getContract({
+    address: SEQUENCE_MARKETPLACE_V1_ADDRESS,
+    abi: Orderbook_ABI,
```
Comment (AlexanderKolberg) [inline src/marketplace/create_listings.ts]: I think we should default to V2 and only use V1 if the user adds an extra argument `--marketplaceVersion 1` or smth

File: src/marketplace/create_listings.ts (hunk 3 above)
Comment (LukasJenicek) [reply src/marketplace/create_listings.ts]: Oh I totally forgot this one.. Thanks for noticing

File: .prettierrc (hunk 1 above)
Comment (LukasJenicek) [reply .prettierrc]: We should propagate this change to all repos then :)

File: src/marketplace/create_listings.ts [hunk 6]
```diff
@@ -1,196 +1,277 @@
...
+  }
+
+  let marketplaceAddress = SEQUENCE_MARKETPLACE_V2_ADDRESS;
+  if (options.marketplaceVersion === 'v1') {
+    marketplaceAddress = SEQUENCE_MARKETPLACE_V1_ADDRESS;
+  }
+
+  let marketplaceABI = SequenceMarktetplace_V2_ABI;
+  if (options.marketplaceVersion == 'v1') {
+    let marketplaceABI = SequenceMarktetplace_V1_ABI;
```
Comment (AlexanderKolberg) [inline src/marketplace/create_listings.ts]:
```suggestion
    marketplaceABI = SequenceMarktetplace_V1_ABI;
```

File: src/marketplace/create_listings.ts [hunk 7]
```diff
@@ -1,196 +1,277 @@
...
+
+  const provider = new ethers.JsonRpcProvider(nodeUrl);
+  const wallet = new ethers.Wallet(privateKey, provider);
+
+  // Set up the wallet client
+  const walletClient = createWalletClient({
+    account: privateKeyToAccount(privateKey),
+    chain: {
+      id: chainConfig.chainId,
+      name: chainConfig.name,
```
Comment (AlexanderKolberg) [inline src/marketplace/create_listings.ts]:
```suggestion
      name: chainConfig.title,
```
viem uses the name to mean a human-readable name, while sequence uses name to mean 'sequence service name'

**Review by AlexanderKolberg — CHANGES_REQUESTED** [review]

File: src/marketplace/create_listings.ts (hunk 6 above)
Comment (LukasJenicek) [reply src/marketplace/create_listings.ts]: ABI is a constant so I had to do it like this: 

```
  const marketplaceABI =
    options.marketplaceVersion == 'v2'
      ? SequenceMarketplace_V2_ABI
      : SequenceMarketplace_V1_ABI;
```

**Review by AlexanderKolberg — APPROVED** [review]
GREAT 🚀

**Pull request merged**